src/build_queue.rs

```diff
@@ -266,13 +266,27 @@ impl BuildQueue {
                 .map(|r| PackageKind::Registry(r.as_str()))
                 .unwrap_or(PackageKind::CratesIo);
 
-            if let Err(err) = builder
+            match builder
                 .update_toolchain()
                 .context("Updating toolchain failed, locking queue")
             {
-                report_error(&err);
-                self.lock()?;
-                return Err(err);
+                Err(err) => {
+                    report_error(&err);
+                    self.lock()?;
+                    return Err(err);
+                }
+                Ok(true) => {
+                    // toolchain has changed, purge caches
+                    if let Err(err) = builder
+                        .purge_caches()
+                        .context("purging rustwide caches failed, locking queue")
```

Comment:
Do we call `update_toolchain()` anywhere else? It might make sense to do the purging inside that function next to `add_essential_files()`?

Reply:
> Do we call update_toolchain() anywhere else? 

Sure, but in manual CLI commands. 

> It might make sense to do the purging inside that function next to add_essential_files()?

While that was an initial thought I had too, this seems to be an unintuitive/surprising place for this cleanup to be in.   IMO no one calling `.update_toolchain` would expect this to clear the rustwide cache folders / cargo caches. 

Generating static files for the toolchain fits in, but cleaning the whole workspace not :)